[FLINK-17073] [checkpointing] checkpointing backpressure if there are too many checkpoints to clean

echauchot opened the pull request:
## What is the purpose of the change

Back pressure the checkpointing system when there are too many checkpoints to clean.
See design doc here: https://docs.google.com/document/d/1q0y0aWlJMoUWNW7jjsM8uWfHsy2dM6YmmcmhpQzgLMA/edit?usp=sharing

Please see comments in the implementation plan section for an explanation of the choices made (classes responsibilities, interfaces, thresholds, resume conditions ...)

## Brief change log

- CheckpointStores are now responsible for checkpoints to clean count. This count is incremented when an asynchronous execution of checkpoint discard is submitted and decremented when the discard is finished (either fail or success) 
- CheckpointRequestDecider stops choosing new checkpoint trigger requests while there are too many checkpoints to clean. When the number of checkpoints to clean becomes below the threshold, CheckpointRequestDecider starts to pick up checkpoint requests again.
- threshold of max checkpoints to clean is configurable as a new configuration item.


## Verifying this change

This change added tests and can be verified as follows: stop and resume checkpoint trigger request pick up by CheckpointRequestDecider.
I did not make a IT test for endurance (verify there is no OOM error) when there is a high frequency of checkpoints cleaning requests. I'm not sure it is needed, tell me if you think it's needed.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes checkpointing)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not applicable)

R: @ifndef-SleePy @rkhachatryan . Thanks guys !

echauchot commented:
@ifndef-SleePy I applied your suggestions of the design doc PTAL.
Also is the test I added enough ?

echauchot commented:
Also I rebased on master because checkpointRequestDecider was changed in master

echauchot commented:
@ifndef-SleePy @rkhachatryan I applied all your comments of the design doc PTAL

echauchot commented:
> Thanks for the PR @echauchot
=> Thanks Roman for the detailed review and also for your suggestions throughout the design ! 
Comments inline and in the reviewed code.
> 
> I think the changes are correct in general, I've left some comments in code.
> 
> Besides that, I have some general remarks:
> 
> 1. Please make sure CI build passes (see @flinkbot comments). I think you missed the license in a new file.
=> Yes I saw the RAT failure in Azure CI build, it did not show up using the verify maven phase when I launched the build locally, that is why I forgot it :)
> 2. Cleanup commit history (see also [changes guide](https://flink.apache.org/contributing/code-style-and-quality-pull-requests.html))
>    remove reverted, e.g. `de8197`
>    squash, e.g. `06c763`
>    make each commit correct, e.g. `numberOfCleaningCheckpointsSupplier.get() > maxQueuedRequests` check in `05f1cb`
>    probably it's easier to have just one commit? I think the changes aren't so big and are related
=> yes I left separate commits to ease the review (allow to isolate parts corresponding to comments in the design doc and isolate further review commits). My plan was indeed to do all the squashing at the end of the review. This can definitely be a single commit as the changes are bound to only one module and do not change the public API etc...
> 3. I think we can re-structure the code so that ZKStore doesn't depend on `CheckpointCleaner`; for this, include a call to `CheckpointCleaner` into `CompletedCheckpoint.discardCallback`
=> how can we do that ? I can decrement the counter in the callback which is executed in the Executor but the incrementation of the counter is done before submitting the callback to the executor, so outside the callback. Do I misunderstand something ?

> 4. These tests are missing:
> 
> * unit tests for `CheckpointsCleaner`
=> it is just a delegate for checkpoint async tasks which are already tested in ZooKeeperCompletedCheckpointStoreTest, PendingCheckpoint. The only thing added is the counter inc/dec. I added a test only for CheckpointRequestDeciderTest to test that when threshold is exceeded, there is no CP request triggered and that there is one when the counter decreases. 
Isn't the UTest the second test below ?
> * test  that checkpoint request is triggered after deletion completes

echauchot commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointsCleaner.java:
```diff
@@ -0,0 +1,86 @@
+ * Delegate class responsible for checkpoints cleaning and counting the number of checkpoints yet
+ * to clean.
+ */
+public class CheckpointsCleaner {
```
sure, that is what I usually do but I forgot.

echauchot commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointsCleaner.java:
```diff
@@ -0,0 +1,86 @@
+public class CheckpointsCleaner {
+	/** The executor used for asynchronous calls, like potentially blocking I/O. */
+	private final Executor executor;
+	AtomicInteger numberOfCheckpointsToClean;
```
ditto

echauchot commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointsCleaner.java:
```diff
@@ -0,0 +1,86 @@
+	private final Executor executor;
+	AtomicInteger numberOfCheckpointsToClean;
+
+	Runnable checkpointCleaningFinishedCallback;
```
totally agree thanks for the suggestion

echauchot commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointsCleaner.java:
```diff
@@ -0,0 +1,86 @@
+	 * @param discardCallback the discard callback to call
+	 */
+	public void asyncDiscardCheckpointAndCountCheckpoints(CompletedCheckpoint completedCheckpoint, ThrowingConsumer<CompletedCheckpoint, Exception> discardCallback){
+		numberOfCheckpointsToClean.incrementAndGet();
```
I meant to call increment before the submission of the callback to the executor because the CP cleaning task is not yet in the executor queue. But I think you're right, I can call it in the executor task (just before the call of accept on the callback)

rkhachatryan replied:
Yes, you are right.

echauchot commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointsCleaner.java:
```diff
@@ -0,0 +1,86 @@
+	public void asyncDiscardPrivateStatesAndCountCheckpoints(
+		Map<OperatorID, OperatorState> operatorStates, CheckpointStorageLocation targetLocation,
+		long checkpointId, JobID jobId) {
+		numberOfCheckpointsToClean.incrementAndGet();
```
ditto

echauchot commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointsCleaner.java:
```diff
@@ -0,0 +1,86 @@
+					checkpointId, jobId, t);
+			} finally {
+				operatorStates.clear();
+				numberOfCheckpointsToClean.decrementAndGet();
```
this method is called in `PendingCP.dispose()` which is called in `PendingCP.finalizeCheckpoint()` which is called when CP are completed. so if I do so, a new CP trigger request will be picked when a given CP is completed and then when it is discarded. So 2 more CP requests for one single checkpoint. Should I make that the counter logic is called only in case of failure (`PendingCP.abort()`) to avoid that case ?

rkhachatryan replied:
`PendingCP.finalizeCheckpoint()` calls `PendingCP.dispose()` with `releaseState=false`, so this method should only run for aborts.
And if it ever changes, I don't think it's a problem to check the queue one more time.

echauchot replied:
ok thanks for this, makes sense

echauchot replied:
done

echauchot commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCheckpointRecoveryFactory.java:
```diff
@@ -44,17 +44,17 @@ public ZooKeeperCheckpointRecoveryFactory(
 			Configuration config,
 			Executor executor) {
-		this.client = checkNotNull(client, "Curator client");
+			this.client = checkNotNull(client, "Curator client");
```
+1 thanks for pointing out !

echauchot replied:
In Beam project, we have a task that reformats the code that we run before submitting a PR. Maybe worth adding such a task to Flink, WDYT ?

rkhachatryan replied:
That would be nice, but current Flink codebase doesn't fully conform to the style guide, so such reformatting would add many unrelated changes.

echauchot replied:
I'm not sure it's a problem with proper community communication (now auto-reformat) and a single reformatting PR with proper commit message. But it's your call.

echauchot commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java:
```diff
@@ -289,7 +290,7 @@ public static CompletedCheckpointStore createCompletedCheckpoints(
 			JobID jobId,
 			int maxNumberOfCheckpointsToRetain,
-			Executor executor) throws Exception {
+			Executor executor, CheckpointsCleaner checkpointsCleaner) throws Exception {
```
+1 thanks for pointing out !

echauchot commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorMasterHooksTest.java:
```diff
@@ -459,7 +460,8 @@ private CheckpointCoordinator instantiateCheckpointCoordinator(
 				new MemoryStateBackend(),
-				Executors.directExecutor(),
+				executor,
+				new CheckpointsCleaner(executor),
```
agree, thanks for the suggestion

echauchot replied:
after refactoring (see below) now that CheckpointsCleaner does nothing (just updates the counter through a callback and expose its value with a getter) I don't think a mock impl still makes sense.

echauchot replied:
with last refactoring: cleaning logic is back in the CheckpointsCleaner. In tests, where we can use a noop checkpointsCleaner I used a noOpCleanCallBack.

echauchot commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java:
```diff
@@ -519,24 +522,8 @@ private void dispose(boolean releaseState) {
-						}
-					});
-
+					checkpointsCleaner.asyncDiscardPrivateStatesAndCountCheckpoints(operatorStates,
```
`CheckpointsCleaner` is just a deledate that submits tasks to the executor as before. Do you mean that the synchronisation is lost because `CheckpointsCleaner` object is created on the `buildGraph()` thread ? Sorry I don't know the threading model of Flink very well yet.

rkhachatryan replied:
I missed that in the previous version it also was asynchronous, so synchronization is fine. Sorry for confusion.

echauchot commented:
@rkhachatryan I answered your comments. PTAL and thanks again for the review

rkhachatryan commented:
Thanks a lot for updating the PR, @echauchot,
I'll take a look.

Meantime, let CI run :)
@flinkbot run azure

echauchot commented:
@rkhachatryan I havent pushed new code yet as I still have interrogations: the one on where to put the increment, the one on double scheduleTriggerRequest call in some cases, the one on synchronisation lost.

echauchot commented:
@rkhachatryan I'm leaving on vacation for 2 weeks this evening, so I could update the PR only when I come back. Nevertheless feel free to tell me what you think about my comments/questions in the meantime.

rkhachatryan commented:
Thanks for the update @echauchot 

I see there are some git conflicts now, can you rebase your branch to the latest master? Besides, there are some checkstyle violations (CI is still failing).


> 3. I think we can re-structure the code so that ZKStore doesn't depend on CheckpointCleaner; for this, include a call to CheckpointCleaner into CompletedCheckpoint.discardCallback
> Ah, you mean calling increment in the executor thread not in the other thread (See #13040 (comment))?

No, I mean there are
1. cyclic dependency (coordinator <-> cleaner through a callback) which makes us use non-final volatile
1. dependency on "different levels" (ZKStore and cleaner) which makes us pass the cleaner while building execution graph.

Let me illustrate both the problem and a possible solution with a diagram:
![Untitled drawing](https://user-images.githubusercontent.com/3939322/90243966-20e69400-de30-11ea-93c2-b35729d3a6e1.png)

> 4.  it is just a delegate 
> Isn't the UTest you want the test that checkpoint request is triggered after deletion completes ?

I don't it's "just a delegate" because it contains this counter and, more importantly a callback. I think these should be covered by tests. And as the whole feature involves multiple components, I think we need an integration test.

5. `PendingCheckpoint.executor` is not used anymore so it can be deleted

Thanks for letting know about your timing. Have a good time :)

echauchot commented:
> Thanks for the update @echauchot
> 
> I see there are some git conflicts now, can you rebase your branch to the latest master? Besides, there are some checkstyle violations (CI is still failing).
> 

sure

> > 1. I think we can re-structure the code so that ZKStore doesn't depend on CheckpointCleaner; for this, include a call to CheckpointCleaner into CompletedCheckpoint.discardCallback
> >    Ah, you mean calling increment in the executor thread not in the other thread (See #13040 (comment))?
> 
> No, I mean there are
> 
> 1. cyclic dependency (coordinator <-> cleaner through a callback) which makes us use non-final volatile
> 2. dependency on "different levels" (ZKStore and cleaner) which makes us pass the cleaner while building execution graph.
> 
> Let me illustrate both the problem and a possible solution with a diagram:
> ![Untitled drawing](https://user-images.githubusercontent.com/3939322/90243966-20e69400-de30-11ea-93c2-b35729d3a6e1.png)
> 

ok thanks for the diagram,  I see your point, I missed this dependency cycle. Such a class diagram on the right would require to move asyncDiscardCheckpointAndCountCheckpoints logic to the checkpoint class itself because the cleaning counter needs to be incremented before the CP discard submission (before try/catch) and the cleaning finished call and decrement need to be done in the finally. That means that the checkpoint cleaner will be responsible only for updating the counter.

> > 1. it is just a delegate
> >    Isn't the UTest you want the test that checkpoint request is triggered after deletion completes ?
> 
> I don't it's "just a delegate" because it contains this counter and, more importantly a callback. I think these should be covered by tests. And as the whole feature involves multiple components, I think we need an integration test.

ok

> 
> 1. `PendingCheckpoint.executor` is not used anymore so it can be deleted
> 

sure

> Thanks for letting know about your timing. Have a good time :)

echauchot commented:
@rkhachatryan I just pushed refactoring code to comply with your suggested diagram. To ease the review I left it as a new commit so that you could see the change. But basically what I did is: 
move checkpoint cleaning logic from _CheckpointCleaner_ to _Checkpoint_ classes to avoid dependencies to _checkpointCleaner_ and make checkpointCleaner only track number of checkpoints to clean. In checkpoint classes interact with checkpoint coordinator and with checkpointCleaner via callbacks. 
There is no more dependency cycle.
I also rebased on master. 
Please take a look and validate the refactoring before I write the test. Also please read my comment in the discussion about the refactoring.

echauchot commented:
@flinkbot run azure

echauchot commented:
@rkhachatryan wait, I'd like to improve something: I created fields in the checkpoint classes for the callbacks and executor, but I prefer passing them as parameters

echauchot commented:
@flinkbot run azure

echauchot commented:
@flinkbot run travis

echauchot commented:
@rkhachatryan aslo please note that I tried improving the refactoring by passing the callbacks where needed instead of storing them as fields of _CompletedCheckpoint_ but that would have led to a dependency to _CheckpointCleaner_ and _CheckpointsCoordinator_ in _ZookeeperCheckpointStore_ so I prefered storing the callbacks inside the checkpoint object at creation time where these deps already exist.

rkhachatryan commented:
Thanks for updating your PR @echauchot 
There is no cyclic dependency now and `CompletedCheckpointStore`s are not affected!

I'm not sure I understood you about having discard logic in `Pending/CompletedCheckpoint` classes. I'd prefer not to put execution (or any) logic there as their concern IMO is to just represent checkpoints.

I think something like this should be possible:
```
public class CheckpointsCleaner {
    public void clean(Runnable cleanAction, Runnable postCleanAction) {
        counter.incrementAndGet();
        executor.execute(() -> {
            try { cleanAction.run(); }
            finally {
                counter.decrementAndGet();
                postCleanAction.run();
            }
        });
    }
}
```
Then in `CheckpointCoordinator`:
```
pendingCheckpoint.finalizeCheckpoint(checkpointsCleaner::clean)
// CompletedCheckpoint saves this callback in a field
```

And in `CompletedCheckpoint`:
```
void asyncDiscardCheckpointAndCountCheckpoint(Consumer<CompletedCheckpoint> discardAction) {
    discardCallbackRunner.accept(
        () -> discardAction.accept(this),
        checkpointCleaningFinishedCallback);
}
```

WDYT?

rkhachatryan commented:
FYI @echauchot, I'll be on vacation until Sep 21.
I'll be available periodically, but if not then I hope @ifndef-SleePy or @pnowojski will review/merge this PR.

echauchot commented:
@rkhachatryan thanks for letting me know. When are you leaving ? I plan to update the PR tomorrow, will it be before your vacation?

rkhachatryan commented:
I'm not sure if I'll be able to look at it tomorrow, but I'm pretty sure it will be merged into the upcoming Flink release anyways.

echauchot commented:
> checkpointCleaningFinishedCallback


Yes I think this refactoring will work from _CompletedCheckpoint_ and is indeed better as it allows to keep the cleaning logic inside _CheckpointsCleaner_. But for _PendingCheckpoint_, there was already, before this PR, some cleaning logic inside _PendingCheckpoint.dispose()_. Do you want that we move this logic to _CheckpointsCleaner_ as well ? 
If yes, it would mean another callback like _CheckpointsCleaner.cleanStates()_

echauchot commented:
> Thanks for updating your PR @echauchot
> There is no cyclic dependency now and `CompletedCheckpointStore`s are not affected!
> 
> I'm not sure I understood you about having discard logic in `Pending/CompletedCheckpoint` classes. I'd prefer not to put execution (or any) logic there as their concern IMO is to just represent checkpoints.
> 
> I think something like this should be possible:
> 
> ```
> public class CheckpointsCleaner {
>     public void clean(Runnable cleanAction, Runnable postCleanAction) {
>         counter.incrementAndGet();
>         executor.execute(() -> {
>             try { cleanAction.run(); }
>             finally {
>                 counter.decrementAndGet();
>                 postCleanAction.run();
>             }
>         });
>     }
> }
> ```
> 
> Then in `CheckpointCoordinator`:
> 
> ```
> pendingCheckpoint.finalizeCheckpoint(checkpointsCleaner::clean)
> // CompletedCheckpoint saves this callback in a field
> ```
> 
> And in `CompletedCheckpoint`:
> 
> ```
> void asyncDiscardCheckpointAndCountCheckpoint(Consumer<CompletedCheckpoint> discardAction) {
>     discardCallbackRunner.accept(
>         () -> discardAction.accept(this),
>         checkpointCleaningFinishedCallback);
> }
> ```
> 
> WDYT?

Yes I think this refactoring will work from _CompletedCheckpoint_ and is indeed better as it allows to keep the cleaning logic inside _CheckpointsCleaner_. But for _PendingCheckpoint_, there was already, before this PR, some cleaning logic inside _PendingCheckpoint.dispose()_. Do you want that we move this logic to _CheckpointsCleaner_ as well ? 
If yes, it would mean another callback like _CheckpointsCleaner.cleanStates()_

echauchot commented:
@ifndef-SleePy or @pnowojski  @rkhachatryan I addressed all the comments PTAL. 
There is just the UTest

echauchot commented:
@ifndef-SleePy or @pnowojski @rkhachatryan I added the integration test, so all comments have been addressed. PTAL. If it is ready to merge, I'll squash commits.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpoint.java:
```diff
@@ -210,6 +221,21 @@ public void registerSharedStatesAfterRestored(SharedStateRegistry sharedStateReg
+	 * and count the number of checkpoints that are waiting to clean.
+	 */
+	void asyncDiscardCheckpointAndCountCheckpoint(ThrowingConsumer<CompletedCheckpoint, Exception> discardCallback, Executor executor){
+		cleanCallback.accept(() -> {
+			try {
+				discardCallback.accept(this);
```
This callback calling callback and `callback.accept(this)` doesn't look very readable to me. 

Looking at `discardCallback`, I see:
1. it's a call to `CompletedCheckpoint.doDiscard` if some condition is `true`
1. the condition can be evaluated early

Therefore, we can get rid of it by:
1. evaluating condition ASAP in `ZooKeeperCompletedCheckpointStore` (maybe extract methods `CompletedCheckpoint.shouldDiscardOnShutdown` and `OnSubsume`)
1. passing `boolean discard` to `tryRemoveCompletedCheckpoint` and further instead of callback
1. replacing `discardCallback.accept(this);`
with `if (discard) this.doDiscard();`. 
in `CompletedCheckpoint.asyncDiscardCheckpointAndCountCheckpoint`:

WDYT?

echauchot replied:
This is the code you gave at the end of [your comment](https://github.com/apache/flink/pull/13040#issuecomment-686639977) I just reused it. But anyway no problem about trying to improve it.

Regarding your proposed refactoring: I understand that in place of passing callbacks `CompletedCheckpoint::discardOnSubsume` and `CompletedCheckpoint::discardOnShutdown`, as they both call `CompletedCheckpoint.doDiscard()` you prefer calling it directly. But it would require to evaluate the conditions inside ZooKeeperStore:

- for `CompletedCheckpoint::discardOnSubsume`: means moving condition `if (props.discardOnSubsumed()) `

- for `CompletedCheckpoint::discardOnShutdown`: means moving conditions `if (jobStatus == JobStatus.FINISHED && props.discardOnJobFinished() ||
				jobStatus == JobStatus.CANCELED && props.discardOnJobCancelled() ||
				jobStatus == JobStatus.FAILED && props.discardOnJobFailed() ||
				jobStatus == JobStatus.SUSPENDED && props.discardOnJobSuspended()) `
=> I'm not sure it's worth it, violating the integrity of CompletedCheckpoint object for the sake a readable `asyncDiscardCheckpointAndCountCheckpoint` method.

rkhachatryan replied:
Sure, we shouldn't break the encapsulation of `CompletedCheckpoint`, but we could extract these conditions into public methods (`boolean shouldDiscardOnXxx()`) which would be used by ZooKeeperStore.
If you think it doesn't worth it I'm also fine with the current approach.

echauchot replied:
I tried this refactoring and I find weird to evaluate the conditions at several places I find it more clear to have conditions evaluated in the `discardOn*` methods and have these methods called as a whole. I prefer this even if it means having unnatural `discardCallback.accept(this);` So if you don't mind I'll leave it as it is.

rkhachatryan commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStoreITCase.java:
```diff
@@ -283,6 +286,54 @@ public void testConcurrentCheckpointOperations() throws Exception {
+		CheckpointRequestDecider checkpointRequestDecider =  new CheckpointRequestDecider(maxCleaningCheckpoints, unused ->{}, clock, 1, new AtomicInteger(0)::get, checkpointsCleaner::getNumberOfCheckpointsToClean);
+
+		final int maxCheckpointsToRetain = 1;
+		Executors.PausableThreadPoolExecutor executor = Executors.pausableExecutor();
```
I think `ManuallyTriggeredScheduledExecutor` would suffice here.
Did you consider using it?

echauchot replied:
I was unaware of this class. Thanks for pointing out ! Yes the public API of this class fits the ITest needs.

rkhachatryan commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStoreITCase.java:
```diff
@@ -283,6 +286,54 @@ public void testConcurrentCheckpointOperations() throws Exception {
+			checkpointStore.addCheckpoint(completedCheckpoint);
+		}
+
+		Thread.sleep(100L); // give time to submit checkpoints for cleaning
+
+		int nbCheckpointsSubmittedForCleaningByCheckpointStore = nbCheckpointsToInject - maxCheckpointsToRetain;
+		assertEquals(nbCheckpointsSubmittedForCleaningByCheckpointStore, checkpointsCleaner.getNumberOfCheckpointsToClean());
```
This `Thread.sleep` can be fragile. How about replacing it with a loop:
```
while (checkpointsCleaner.getNumberOfCheckpointsToClean() < nbCheckpointsToInject - maxCheckpointsToRetain) {
    Thread.sleep(10);
}
```
?

(ditto: the next `sleep`)

echauchot replied:
agree. I don't like absolute sleeps either but I did not go into so much thinking about that, I first wanted that the ITest behavior is validated before dealing with details.

echauchot replied:
Some thing troubles me though: using the thing we want to test (`checkpointsCleaner.getNumberOfCheckpointsToClean()`) as a condition for a waiting loop seems not a good idea to me as if the tested element fails, we could have an infinite loop in place of a failing ITest. Maybe more trying to synchronise the ITest and checkpointsStore submit somehow (even by raising the visibility of some internals for the purpose of test). 

WDYT ?

echauchot replied:
No, I used a CountDownLatch for synchronising the executor and the test

rkhachatryan replied:
I find `CountDownLatch` approach worse compared to `while` loops because:
1. it's more fragile because ZKStore can submit several `Runnable`s
1. I need to follow when latches are triggered
1. with a new Executor added, it's more complex

For 1, we should probably use `triggerAll` instead of `trigger`. But still it's unclear what should be the counts in latches.
For 3, I think executor could provide methods like `awaitSubmitted(count)` and `awaitAllTriggered`.

But I'd still prefer a loop.

(please unresolve this thread)

echauchot replied:
For 1. you're right, `nbCheckpointsSubmittedForCleaning == 2` so `CountDownLatch(1)` is not correct. +1 on `triggerAll`

For 2. true
For 3. it is just an extend with 2 overrides it does not seem a big deal to me. Moreover, this new class is only in scope test.

Still I don't like looping on the condition that we are actually testing (the count of cleaning CP). IMHO, failing with a timeout if the tested count is incorrect is misleading and makes the build longer so IMHO it is not acceptable. 
I'll try to find a better way to sync the executor and the test rather than the CountDownLatch.

PS: I resolved the discussion by mistake. I just unresolved it.

echauchot replied:
@rkhachatryan I used waiting loops instead of `countDownLatch` to synchronise the test and the executor. And I simply used the executor queue size as the condition for the waiting loop. That way we no more have the condition I did not like and we no more have the countDownLatch and the new executor class you did not like. 
Do you find this solution acceptable ?

rkhachatryan replied:
I think the assumption that each cleanup action is a separate runnable (and that there are no other runnable) can break at some point and is less straightforward than `checkpointsCleaner.getNumberOfCheckpointsToClean() < expected`.

echauchot replied:
Sure the executor can be used for other things that checkpointing in a regular pipeline for the general case. But in the case of this test there is no real pipeline, no `CheckpointCoordinator` just a `ZookeeperCheckpointStore` that gets added some checkpoints. `ZookeeperCheckpointStore` uses the executor only for checkpoints cleaning. In such conditions, I'm not sure that the assumption that there is no other runnables than checkpoint cleanup runnables for the executor does not seem that weak to me.
Do you see a case in which this test environment (only a ZookeeperCPStore created) can generate other runnables for the executor ?

rkhachatryan replied:
Currently not, but I mean possible future changes that can happen in `ZooKeeperCompletedCheckpointStore` or `CheckpointsCleaner`.

echauchot replied:
yes but we are talking about test code not production code. That is not a big deal if a test starts to fail when production code is changed. If that happens, the author will change the test code. Can we merge the PR ?

rkhachatryan replied:
I think forcing the developer to change the tests should be avoided if possible, and here it is possible. 
Also, condition on the public interface rather than internal details makes expectations more clear.

From my side, the PR can be merged after resolving this issue and grooming commits.

Probably some other developers would like to take a brief look - I'll ask about it.

echauchot replied:
I mean, a test reflects the state of the production code at a given time, and at this time a checkpoint cleaning task is submitted as a single runnable. Anyway, if the problem is relying on internals of the executor, I can add await methods in the public API of the executor as you suggested in a previous comment. That way the test author will know that the test expects a certain number of submitted and triggered runnables. I pushed a commit in that sense. Is this ok for you ?

rkhachatryan replied:
The latest commit still makes the assumption about the number of Runnables.
I just dealt with a similar assumption in another test while rebasing my own PR, and would like to avoid it in the future :) 
But let's maybe wait for other developer's feedback. I asked one of my colleagues to take a look at your PR.

echauchot replied:
Yes let's wait for other feedback @tillrohrmann maybe as he knows the ticket very well ?
I also wanted to mention that the aim of this ticket is to count the runnables stacked in the executor (before getting executed) so that they do not lead to OOM. So, it does not choc me to rely on the count of runnables in the test.

echauchot replied:
By the way, we will need another reviewer anyway to merge the PR.

tillrohrmann replied:
I am not entirely sure about which current code snippet we are talking because this one seems to point at some outdated version. Hence I can only give my general thoughts: I think that Roman has a valid point that relying on implementation details such as the number of submitted `Runnables` is more brittle than relying on some public API. I myself have built quite some of these tests and regretted this shortly after when having to refactor these tests. Hence, my recommendation would be to avoid this as much as possible and rather try to rely on the public API of the components under test.

echauchot replied:
Hi @tillrohrmann thanks for your comment ! 

Regarding the code version, we are referring to the last commit (https://github.com/apache/flink/pull/13040/commits/aaf4aca0d5ce666001035f6ae66bca841f87c74d). 

We agreed on using a waiting loop to sync the executor and the test, the only point of discussion is on the condition of this waiting loop: I did not want to use the element that this test tests (the number of cleaning checkpoints) as a condition for a waiting loop. I agree that using internals for this condition is not ideal but do you have any condition to suggest for these waiting loops other than the number of cleaning checkpoints or the number of runnables submitted ?

tillrohrmann replied:
No, I think waiting on the number of checkpoints to clean would be the best option as far as I know. If you are afraid of indefinite waiting you could also use `CommonTestUtils.waitUntilCondition` which allows to specify a timeout for the check.

echauchot replied:
Fair enough, if the condition on internals is too risky and if the condition on nb of checkpoints to clean is the only other solution, I'll do that. Thanks for your opinion guys !

rkhachatryan reviewed (changes requested):
Thanks a lot for updating your PR @echauchot. Most changes LGTM.
I've left some comments about the callbacks and the ITCase, please take a look.
And please also fix the build failures in AZP.

echauchot commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointsCleaner.java:
```diff
@@ -0,0 +1,79 @@
+			// discard the private states.
+			// unregistered shared states are still considered private at this point.
+			try {
+				StateUtil.bestEffortDiscardAllStateObjects(operatorStates.values());
+				targetLocation.disposeOnFailure();
```
Yes that is what I asked in [this](https://github.com/apache/flink/pull/13040#issuecomment-688845131) comment if I needed to move the cleaning logic from `PendingCheckpoint` to `CheckpointsCleaner` but, as you were off, I took the choice to move it to have all cleaning logic at the same place. Anyway, I could move it back to `PendingCheckpoint`. 

For that, we could indeed use `cleanCheckpoint`  for doing the `cleanStates()` job and call				`operatorStates.clear(); CheckpointCoordinator::scheduleTriggerRequest();` as a `postCleanAction` callback. That way we have only `cleanCheckpoint()` in `CheckpointsCleaner`. Is that good for you ? (I would like it to be the final refactoring :) )

echauchot replied:
@rkhachatryan can you confirm that this is indeed the refactoring that you want before I code it ?

rkhachatryan replied:
Sorry, my reply got lost for some reason.

I think that `operatorStates.clear();` should remain in `PendingCheckpoint` too lest break encapsulation.
So the original Runnable can be submitted as `cleanAction` to `CheckpointCleaner.cleanCheckpoint`.:
```
try {
    StateUtil.bestEffortDiscardAllStateObjects(operatorStates.values());
    targetLocation.disposeOnFailure();
} catch (Throwable t) {
    LOG.warn("Could not properly dispose the private states in the pending checkpoint {} of job {}.",
        checkpointId, jobId, t);
} finally {
    operatorStates.clear();
}
```

echauchot replied:
Fair enough

echauchot commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpoint.java:
```diff
@@ -108,6 +110,11 @@
 	@Nullable
 	private transient volatile CompletedCheckpointStats.DiscardCallback discardCallback;
 
+	private final CheckpointsCleaningRunner cleanCallback;
+
+	private final SerializableRunnable checkpointCleaningFinishedCallback;
```
sure, thx for pointing out !

echauchot commented:
@rkhachatryan I addressed your comments PTAL

echauchot commented:
@flinkbot run azure

echauchot commented:
@rkhachatryan @tillrohrmann I'll be slightly off for the next 2 days as I'm attending the apachecon and also giving a talk.

echauchot commented:
@rkhachatryan @tillrohrmann I removed the last commit and put a new one changing the waiting condition. PTAL.

echauchot commented:
@flinkbot run azure

rkhachatryan commented:
Thanks for updating your PR @echauchot 
Can you please squash related commits and rebase to the latest master?

echauchot commented:
@rkhachatryan sure, I'll do that after the conference

echauchot commented:
@tillrohrmann I rebased on master and squashed the PR to only 2 commits:
- production code + UTest of the CheckpointsDecider
- integration test
Can we merge the PR ?

rkhachatryan commented:
Hey @echauchot, thanks a lot for the contribution. The changes look mostly good to me, I will take care of some follow up clean ups and merge it in a couple of days.

echauchot commented:
Hey @rkhachatryan I thought you were not a flink committer and you could not merge the PR ?

pnowojski commented:
@echauchot it was a mental shortcut :) (I will be doing the final merge of your contribution and the follow up refactor that @rkhachatryan is doing).

pnowojski commented:
Merged to master as a part of https://github.com/apache/flink/pull/13565

rkhachatryan commented:
Thanks for this effort @echauchot and for merging @pnowojski
